### PR TITLE
KeyspaceStatus: Include Partitioning Readiness

### PR DESCRIPTION
**Pull request by PrismaPhonic**
This PR expands the partitioning status section to include readiness information and provides a high-level condition to assess whether the partitioning that is serving writes has a ready tablet count that matches its desired tablet count.

### CODE REVIEW

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (avaidyanatha) [inline pkg/apis/planetscale/v2/vitesskeyspace_types.go:421]: Nit: The zero value is 0, so this is fine right?
```suggestion
	var desiredTablets int32
```

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (avaidyanatha) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:202]: My initial thought was that this looks copypasta-ish and all of this logic could sit in one subroutine or method? So you could call something like `updateTabletCounts()` to set all the values in one place.

Either way, this isn't that important to me and I won't block on it, just let me know what you feel about it.

**Review by avaidyanatha — APPROVED** [review]
LGTM, two thoughts.

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (PrismaPhonic) [reply pkg/apis/planetscale/v2/vitesskeyspace_types.go:421]: I was copying logic from another constructor function above in the same file. But I believe you're right on that.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go:202]: That's true to some extent but would require passing in the thing we want to mutate. I tend to avoid doing that because it can lead to buggy code that's unclear. I find it better usually to design pure functions that don't mutate data and return aggregated results. I agree it's less efficient, but I don't personally believe that two functions that do similar things are always bad.

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (PrismaPhonic) [reply pkg/apis/planetscale/v2/vitesskeyspace_types.go]: Updated here, and also updated the place I copied it from.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (avaidyanatha) [reply pkg/controller/vitesskeyspace/reconcile_shards.go:202]: I'm totally cool with keeping it how it is, just wanted to throw that option out there just in case you had thought of it yourself and were on the edge of making the change or not :)

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (avaidyanatha) [reply pkg/apis/planetscale/v2/vitesskeyspace_types.go]: Cool, thanks!

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (enisoc) [inline pkg/apis/planetscale/v2/vitesskeyspace_types.go:465]:
```suggestion
	// VitessKeyspaceReady indicates whether the tablet Pods of the keyspace's serving partitioning are all Ready.
```

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:202]:
```suggestion
func totalReadyTablets(shardNames []string, shardStatus map[string]planetscalev2.VitessKeyspaceShardStatus) int32 {
```

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:162]: We don't need to update the condition message whenever the particular values change. The status values themselves give that info. Also "serving tablets" seems misleading; what we really mean is that the tablets are in the partitioning that's supposed to be serving, but the tablets themselves are not serving (because they aren't ready). Let's just keep the Reason string simple.

```suggestion
				r.setConditionStatus(planetscalev2.VitessKeyspaceReady, corev1.ConditionFalse, "TabletsNotReady", "Some tablet Pods for the serving partitioning are not Ready.")
```

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:164]:
```suggestion
				r.setConditionStatus(planetscalev2.VitessKeyspaceReady, corev1.ConditionTrue, "TabletsReady", "All tablet Pods for the serving partitioning are Ready")
```

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:160]: If we get through the loop without finding any serving partitioning, we should set the Ready condition status to False with reason NoServingPartitioning. Otherwise it'll be left at Unknown/ReconcileFailed which doesn't represent what we actually know in that case.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:206]: This doesn't need to be special case. The loop will just iterate 0 times.

File: pkg/apis/planetscale/v2/vitesskeyspace_types.go
Comment (PrismaPhonic) [reply pkg/apis/planetscale/v2/vitesskeyspace_types.go]: Done.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go]: Done.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go]: Done.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go]: Done.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go:161]: Done.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go]: Removed from all.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (enisoc) [inline pkg/controller/vitesskeyspace/reconcile_shards.go:238]: I think we should look at DesiredTablets so we don't think everything is fine if a tablet is missing. It looks like I got this wrong myself in `vitesscluster/reconcile_keyspaces.go`. Can you fix it there too?

```suggestion
		if shard.ReadyTablets == shard.DesiredTablets {
```

**Review by enisoc — APPROVED** [review]
LGTM if you agree on the one minor correction. Conditionally approved.

File: pkg/controller/vitesskeyspace/reconcile_shards.go
Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go:238]: Good point. I'll fix it in both spots!

Comment (PrismaPhonic) [reply pkg/controller/vitesskeyspace/reconcile_shards.go]: Done.

**Pull request merged**
